File: app/src/main/java/com/pimenta/bestv/feature/workdetail/ui/WorkDetailsFragment.java
/*
 * Copyright (C) 2018 Marcus Pimenta
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

package com.pimenta.bestv.feature.workdetail.ui;

import android.content.ActivityNotFoundException;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.net.Uri;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v17.leanback.app.DetailsSupportFragmentBackgroundController;
import android.support.v17.leanback.widget.Action;
import android.support.v17.leanback.widget.ArrayObjectAdapter;
import android.support.v17.leanback.widget.ClassPresenterSelector;
import android.support.v17.leanback.widget.DetailsOverviewRow;
import android.support.v17.leanback.widget.FullWidthDetailsOverviewRowPresenter;
import android.support.v17.leanback.widget.FullWidthDetailsOverviewSharedElementHelper;
import android.support.v17.leanback.widget.HeaderItem;
import android.support.v17.leanback.widget.ImageCardView;
import android.support.v17.leanback.widget.ListRow;
import android.support.v17.leanback.widget.ListRowPresenter;
import android.support.v17.leanback.widget.OnItemViewClickedListener;
import android.support.v17.leanback.widget.OnItemViewSelectedListener;
import android.support.v17.leanback.widget.Presenter;
import android.support.v17.leanback.widget.Row;
import android.support.v17.leanback.widget.RowPresenter;
import android.support.v4.app.ActivityOptionsCompat;
import android.util.Log;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.pimenta.bestv.BesTV;
import com.pimenta.bestv.BuildConfig;
import com.pimenta.bestv.R;
import com.pimenta.bestv.feature.base.BaseDetailsFragment;
import com.pimenta.bestv.feature.castdetail.ui.CastDetailsActivity;
import com.pimenta.bestv.feature.castdetail.ui.CastDetailsFragment;
import com.pimenta.bestv.feature.widget.render.WorkDetailsDescriptionRender;
import com.pimenta.bestv.feature.widget.render.CastCardRender;
import com.pimenta.bestv.feature.widget.render.VideoCardRender;
import com.pimenta.bestv.feature.widget.render.WorkCardRenderer;
import com.pimenta.bestv.feature.workdetail.presenter.WorkDetailsPresenter;
import com.pimenta.bestv.repository.entity.Cast;
import com.pimenta.bestv.repository.entity.Video;
import com.pimenta.bestv.repository.entity.Work;

import java.util.List;

import javax.inject.Inject;

/**
 * Created by marcus on 07-02-2018.
 */
public class WorkDetailsFragment extends BaseDetailsFragment implements WorkDetailsPresenter.WorkDetailsView {

    public static final String TAG = "WorkDetailsFragment";
    public static final String SHARED_ELEMENT_NAME = "SHARED_ELEMENT_NAME";
    public static final String WORK = "WORK";

    private static final int ACTION_FAVORITE = 1;
    private static final int ACTION_VIDEOS = 2;
    private static final int ACTION_CAST = 3;
    private static final int ACTION_RECOMMENDED = 4;
    private static final int ACTION_SIMILAR = 5;
    private static final int VIDEO_HEADER_ID = 1;
    private static final int RECOMMENDED_HEADER_ID = 2;
    private static final int SIMILAR_HEADER_ID = 3;
    private static final int CAST_HEAD_ID = 4;

    private Action mFavoriteAction;
    private ArrayObjectAdapter mAdapter;
    private ArrayObjectAdapter mActionAdapter;
    private ArrayObjectAdapter mVideoRowAdapter;
    private ArrayObjectAdapter mCastRowAdapter;
    private ArrayObjectAdapter mRecommendedRowAdapter;
    private ArrayObjectAdapter mSimilarRowAdapter;
    private ClassPresenterSelector mPresenterSelector;
    private DetailsOverviewRow mDetailsOverviewRow;
    private DetailsSupportFragmentBackgroundController mDetailsBackground;

    private Work mWork;

    @Inject
    WorkDetailsPresenter mPresenter;

    public static WorkDetailsFragment newInstance() {
        return new WorkDetailsFragment();
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        BesTV.applicationComponent.inject(this);
        mPresenter.register(this);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (mWork == null) {
            mWork = (Work) getActivity().getIntent().getSerializableExtra(WORK);
        }

        setupDetailsOverviewRow();
        setupDetailsOverviewRowPresenter();
        setAdapter(mAdapter);
    }

    @Override
    public void onViewCreated(android.view.View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        mFavoriteAction = new Action(ACTION_FAVORITE, mPresenter.isFavorite(mWork) ? getResources().getString(R.string.remove_favorites) :
                getResources().getString(R.string.save_favorites));
        mActionAdapter.add(mFavoriteAction);
        mPresenter.loadDataByWork(mWork);
    }

    @Override
    public void onDetach() {
        mPresenter.unRegister();
        super.onDetach();
    }

    @Override
    public void onResultSetFavoriteMovie(boolean success) {
        if (success) {
            mFavoriteAction.setLabel1(mWork.isFavorite() ? getResources().getString(R.string.remove_favorites) : getResources().getString(R.string.save_favorites));
            mActionAdapter.notifyItemRangeChanged(mActionAdapter.indexOf(mFavoriteAction), 1);
        }
    }

    @Override
    public void onDataLoaded(List<Cast> casts, List<? extends Work> recommendedWorks, List<? extends Work> similarWorks, List<? extends Video> videos) {
        if (videos != null && videos.size() > 0) {
            mActionAdapter.add(new Action(ACTION_VIDEOS, getResources().getString(R.string.videos)));
            VideoCardRender videoCardRender = new VideoCardRender();
            mVideoRowAdapter = new ArrayObjectAdapter(videoCardRender);
            mVideoRowAdapter.addAll(0, videos);
            final HeaderItem recommendedHeader = new HeaderItem(VIDEO_HEADER_ID, getString(R.string.videos));
            mAdapter.add(new ListRow(recommendedHeader, mVideoRowAdapter));
        }

        if (casts != null && casts.size() > 0) {
            mActionAdapter.add(new Action(ACTION_CAST, getResources().getString(R.string.cast)));
            CastCardRender castCardRender = new CastCardRender();
            mCastRowAdapter = new ArrayObjectAdapter(castCardRender);
            mCastRowAdapter.addAll(0, casts);
            final HeaderItem header = new HeaderItem(CAST_HEAD_ID, getString(R.string.cast));
            mAdapter.add(new ListRow(header, mCastRowAdapter));
        }

        if (recommendedWorks != null && recommendedWorks.size() > 0) {
            mActionAdapter.add(new Action(ACTION_RECOMMENDED, getResources().getString(R.string.recommended)));
            WorkCardRenderer workCardRenderer = new WorkCardRenderer();
            mRecommendedRowAdapter = new ArrayObjectAdapter(workCardRenderer);
            mRecommendedRowAdapter.addAll(0, recommendedWorks);
            final HeaderItem recommendedHeader = new HeaderItem(RECOMMENDED_HEADER_ID, getString(R.string.recommended));
            mAdapter.add(new ListRow(recommendedHeader, mRecommendedRowAdapter));
        }

        if (similarWorks != null && similarWorks.size() > 0) {
            mActionAdapter.add(new Action(ACTION_SIMILAR, getResources().getString(R.string.similar)));
            WorkCardRenderer workCardRenderer = new WorkCardRenderer();
            mSimilarRowAdapter = new ArrayObjectAdapter(workCardRenderer);
            mSimilarRowAdapter.addAll(0, similarWorks);
            final HeaderItem similarHeader = new HeaderItem(SIMILAR_HEADER_ID, getString(R.string.similar));
            mAdapter.add(new ListRow(similarHeader, mSimilarRowAdapter));
        }
    }

    @Override
    public void onRecommendationLoaded(List<? extends Work> works) {
        if (works != null) {
            for (final Work work : works) {
                if (mRecommendedRowAdapter.indexOf(work) == -1) {
                    mRecommendedRowAdapter.add(work);
                }
            }
        }
    }

    @Override
    public void onSimilarLoaded(List<? extends Work> works) {
        if (works != null) {
            for (final Work work : works) {
                if (mSimilarRowAdapter.indexOf(work) == -1) {
                    mSimilarRowAdapter.add(work);
                }
            }
        }
    }

    @Override
    public void onCardImageLoaded(Drawable resource) {
        mDetailsOverviewRow.setImageDrawable(resource);
        mAdapter.notifyArrayItemRangeChanged(0, mAdapter.size());
        setupBackgroundImage();
    }

    @Override
    public void onBackdropImageLoaded(Bitmap bitmap) {
        mDetailsBackground.setCoverBitmap(bitmap);
        mAdapter.notifyArrayItemRangeChanged(0, mAdapter.size());
    }

    private void setupDetailsOverviewRow() {
        mPresenterSelector = new ClassPresenterSelector();
        mPresenterSelector.addClassPresenter(ListRow.class, new ListRowPresenter());
        mAdapter = new ArrayObjectAdapter(mPresenterSelector);

        mDetailsOverviewRow = new DetailsOverviewRow(mWork);
        mPresenter.loadCardImage(mWork);

        mActionAdapter = new ArrayObjectAdapter();
        mDetailsOverviewRow.setActionsAdapter(mActionAdapter);
        mAdapter.add(mDetailsOverviewRow);

        setOnItemViewSelectedListener(new ItemViewSelectedListener());
        setOnItemViewClickedListener(new ItemViewClickedListener());
    }

    private void setupDetailsOverviewRowPresenter() {
        // Set detail background.
        final FullWidthDetailsOverviewRowPresenter detailsPresenter = new FullWidthDetailsOverviewRowPresenter(new WorkDetailsDescriptionRender()) {

            private ImageView mDetailsImageView;

            @Override
            protected RowPresenter.ViewHolder createRowViewHolder(ViewGroup parent) {
                RowPresenter.ViewHolder viewHolder = super.createRowViewHolder(parent);
                mDetailsImageView = viewHolder.view.findViewById(R.id.details_overview_image);
                ViewGroup.LayoutParams lp = mDetailsImageView.getLayoutParams();
                lp.width = getResources().getDimensionPixelSize(R.dimen.movie_width);
                lp.height = getResources().getDimensionPixelSize(R.dimen.movie_height);
                mDetailsImageView.setLayoutParams(lp);
                return viewHolder;
            }
        };
        detailsPresenter.setActionsBackgroundColor(getResources().getColor(R.color.detail_view_actionbar_background, getActivity().getTheme()));
        detailsPresenter.setBackgroundColor(getResources().getColor(R.color.detail_view_background, getActivity().getTheme()));

        // Hook up transition element.
        final FullWidthDetailsOverviewSharedElementHelper sharedElementHelper = new FullWidthDetailsOverviewSharedElementHelper();
        sharedElementHelper.setSharedElementEnterTransition(getActivity(), SHARED_ELEMENT_NAME);
        detailsPresenter.setListener(sharedElementHelper);
        detailsPresenter.setParticipatingEntranceTransition(true);
        detailsPresenter.setOnActionClickedListener(action -> {
            int position = 0;
            switch ((int) action.getId()) {
                case ACTION_FAVORITE:
                    mPresenter.setFavorite(mWork);
                    break;
                case ACTION_SIMILAR:
                    if (mSimilarRowAdapter != null && mSimilarRowAdapter.size() > 0) {
                        position++;
                    }
                case ACTION_RECOMMENDED:
                    if (mRecommendedRowAdapter != null && mRecommendedRowAdapter.size() > 0) {
                        position++;
                    }
                case ACTION_CAST:
                    if (mCastRowAdapter != null && mCastRowAdapter.size() > 0) {
                        position++;
                    }
                case ACTION_VIDEOS:
                    if (mVideoRowAdapter != null && mVideoRowAdapter.size() > 0) {
                        position++;
                    }
                    setSelectedPosition(position);
            }
        });
        mPresenterSelector.addClassPresenter(DetailsOverviewRow.class, detailsPresenter);
    }

    private void setupBackgroundImage() {
        mDetailsBackground = new DetailsSupportFragmentBackgroundController(this);
        mDetailsBackground.enableParallax();
        mPresenter.loadBackdropImage(mWork);
    }

    private final class ItemViewSelectedListener implements OnItemViewSelectedListener {

        @Override
        public void onItemSelected(Presenter.ViewHolder itemViewHolder, Object item, RowPresenter.ViewHolder rowViewHolder, Row row) {
            if (row != null && row.getHeaderItem() != null) {
                switch ((int) row.getHeaderItem().getId()) {
                    case RECOMMENDED_HEADER_ID:
                        Work recommendedWork = (Work) item;
                        if (mRecommendedRowAdapter.indexOf(recommendedWork) >= mRecommendedRowAdapter.size() - 1) {
                            mPresenter.loadRecommendationByWork(mWork);
                        }
                        break;
                    case SIMILAR_HEADER_ID:
                        Work similarWork = (Work) item;
                        if (mSimilarRowAdapter.indexOf(similarWork) >= mSimilarRowAdapter.size() - 1) {
                            mPresenter.loadSimilarByWork(mWork);
                        }
                        break;
                }
            }
        }
    }

    private final class ItemViewClickedListener implements OnItemViewClickedListener {

        @Override
        public void onItemClicked(Presenter.ViewHolder itemViewHolder, Object item, RowPresenter.ViewHolder rowViewHolder, Row row) {
            if (row != null && row.getHeaderItem() != null) {
                switch ((int) row.getHeaderItem().getId()) {
                    case CAST_HEAD_ID:
                        Cast cast = (Cast) item;
                        Bundle castBundle = ActivityOptionsCompat.makeSceneTransitionAnimation(getActivity(),
                                ((ImageCardView) itemViewHolder.view).getMainImageView(), CastDetailsFragment.SHARED_ELEMENT_NAME).toBundle();
                        startActivity(CastDetailsActivity.Companion.newInstance(getContext(), cast), castBundle);
                        break;
                    case VIDEO_HEADER_ID:
                        Video video = (Video) item;
                        Intent intent = new Intent(Intent.ACTION_VIEW,
                                Uri.parse(String.format(BuildConfig.YOUTUBE_BASE_URL, video.getKey())));
                        try {
                            startActivity(intent);
                        } catch (ActivityNotFoundException e) {
                            Log.e(TAG, "Failed to play a video", e);
                        }
                        break;
                    case RECOMMENDED_HEADER_ID:
                    case SIMILAR_HEADER_ID:
                        Work work = (Work) item;
                        Bundle bundle = ActivityOptionsCompat.makeSceneTransitionAnimation(getActivity(),
                                ((ImageCardView) itemViewHolder.view).getMainImageView(), SHARED_ELEMENT_NAME).toBundle();
                        startActivity(WorkDetailsActivity.newInstance(getContext(), work), bundle);
                        break;
                }
            }
        }
    }
}

File: app/src/main/java/com/pimenta/bestv/feature/workdetail/ui/WorkDetailsFragment.kt
/*
 * Copyright (C) 2018 Marcus Pimenta
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */

package com.pimenta.bestv.feature.workdetail.ui

import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.net.Uri
import android.os.Bundle
import androidx.leanback.app.DetailsSupportFragmentBackgroundController
import androidx.leanback.widget.*
import androidx.core.app.ActivityOptionsCompat
import android.view.ViewGroup
import android.widget.ImageView
import com.pimenta.bestv.BesTV
import com.pimenta.bestv.BuildConfig
import com.pimenta.bestv.R
import com.pimenta.bestv.feature.base.BaseDetailsFragment
import com.pimenta.bestv.feature.castdetail.ui.CastDetailsActivity
import com.pimenta.bestv.feature.castdetail.ui.CastDetailsFragment
import com.pimenta.bestv.widget.render.CastCardRender
import com.pimenta.bestv.widget.render.VideoCardRender
import com.pimenta.bestv.widget.render.WorkCardRenderer
import com.pimenta.bestv.widget.render.WorkDetailsDescriptionRender
import com.pimenta.bestv.feature.workdetail.presenter.WorkDetailsPresenter
import com.pimenta.bestv.repository.entity.Cast
import com.pimenta.bestv.repository.entity.Video
import com.pimenta.bestv.repository.entity.Work
import timber.log.Timber
import javax.inject.Inject

/**
 * Created by marcus on 07-02-2018.
 */
class WorkDetailsFragment : BaseDetailsFragment(), WorkDetailsPresenter.View {

    private val actionAdapter: ArrayObjectAdapter by lazy { ArrayObjectAdapter() }
    private val videoRowAdapter: ArrayObjectAdapter by lazy { ArrayObjectAdapter(VideoCardRender()) }
    private val castRowAdapter: ArrayObjectAdapter by lazy { ArrayObjectAdapter(CastCardRender()) }
    private val recommendedRowAdapter: ArrayObjectAdapter by lazy { ArrayObjectAdapter(WorkCardRenderer()) }
    private val similarRowAdapter: ArrayObjectAdapter  by lazy { ArrayObjectAdapter(WorkCardRenderer()) }
    private val mainAdapter: ArrayObjectAdapter by lazy { ArrayObjectAdapter(presenterSelector) }
    private val presenterSelector: ClassPresenterSelector by lazy {
        ClassPresenterSelector().apply {
            addClassPresenter(ListRow::class.java, ListRowPresenter())
        }
    }
    private val detailsBackground: DetailsSupportFragmentBackgroundController by lazy {
        DetailsSupportFragmentBackgroundController(this).apply {
            enableParallax()
        }
    }

    private lateinit var favoriteAction: Action
    private lateinit var detailsOverviewRow: DetailsOverviewRow
    private lateinit var work: Work

    @Inject
    lateinit var presenter: WorkDetailsPresenter

    override fun onAttach(context: Context) {
        super.onAttach(context)
        BesTV.applicationComponent.getWorkDetailsFragmentComponent()
                .view(this)
                .build()
                .inject(this)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        activity?.let {
            work = it.intent.getSerializableExtra(WORK) as Work
        }

        setupDetailsOverviewRow()
        setupDetailsOverviewRowPresenter()
        adapter = mainAdapter
    }

    override fun onViewCreated(view: android.view.View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        favoriteAction = Action(
                ACTION_FAVORITE.toLong(),
                if (presenter.isFavorite(work))
                    resources.getString(R.string.remove_favorites)
                else
                    resources.getString(R.string.save_favorites)
        )
        actionAdapter.add(favoriteAction)
        presenter.loadDataByWork(work)
    }

    override fun onDetach() {
        presenter.dispose()
        super.onDetach()
    }

    override fun onResultSetFavoriteMovie(success: Boolean) {
        if (success) {
            favoriteAction.label1 = if (work.isFavorite) resources.getString(R.string.remove_favorites) else resources.getString(R.string.save_favorites)
            actionAdapter.notifyItemRangeChanged(actionAdapter.indexOf(favoriteAction), 1)
        }
    }

    override fun onDataLoaded(casts: List<Cast>?, recommendedWorks: List<Work>?, similarWorks: List<Work>?, videos: List<Video>?) {
        videos?.let {
            if (it.isNotEmpty()) {
                actionAdapter.add(Action(ACTION_VIDEOS.toLong(), resources.getString(R.string.videos)))
                videoRowAdapter.addAll(0, it)
                mainAdapter.add(ListRow(HeaderItem(VIDEO_HEADER_ID.toLong(), getString(R.string.videos)), videoRowAdapter))
            }
        }

        casts?.let {
            if (it.isNotEmpty()) {
                actionAdapter.add(Action(ACTION_CAST.toLong(), resources.getString(R.string.cast)))
                castRowAdapter.addAll(0, it)
                mainAdapter.add(ListRow(HeaderItem(CAST_HEAD_ID.toLong(), getString(R.string.cast)), castRowAdapter))
            }
        }

        recommendedWorks?.let {
            if (it.isNotEmpty()) {
                actionAdapter.add(Action(ACTION_RECOMMENDED.toLong(), resources.getString(R.string.recommended)))
                recommendedRowAdapter.addAll(0, it)
                mainAdapter.add(ListRow(HeaderItem(RECOMMENDED_HEADER_ID.toLong(), getString(R.string.recommended)), recommendedRowAdapter))
            }
        }

        similarWorks?.let {
            if (it.isNotEmpty()) {
                actionAdapter.add(Action(ACTION_SIMILAR.toLong(), resources.getString(R.string.similar)))
                similarRowAdapter.addAll(0, it)
                mainAdapter.add(ListRow(HeaderItem(SIMILAR_HEADER_ID.toLong(), getString(R.string.similar)), similarRowAdapter))
            }
        }
    }

    override fun onRecommendationLoaded(works: List<Work>?) {
        works?.forEach { work ->
            if (recommendedRowAdapter.indexOf(work) == -1) {
                recommendedRowAdapter.add(work)
            }
        }
    }

    override fun onSimilarLoaded(works: List<Work>?) {
        works?.forEach { work ->
            if (similarRowAdapter.indexOf(work) == -1) {
                similarRowAdapter.add(work)
            }
        }
    }

    override fun onCardImageLoaded(resource: Drawable?) {
        detailsOverviewRow.imageDrawable = resource
        mainAdapter.notifyArrayItemRangeChanged(0, mainAdapter.size())
        presenter.loadBackdropImage(work)
    }

    override fun onBackdropImageLoaded(bitmap: Bitmap?) {
        detailsBackground.coverBitmap = bitmap
        mainAdapter.notifyArrayItemRangeChanged(0, mainAdapter.size())
    }

    private fun setupDetailsOverviewRow() {
        detailsOverviewRow = DetailsOverviewRow(work)
        presenter.loadCardImage(work)

        detailsOverviewRow.actionsAdapter = actionAdapter
        mainAdapter.add(detailsOverviewRow)

        setOnItemViewSelectedListener(ItemViewSelectedListener())
        onItemViewClickedListener = ItemViewClickedListener()
    }

    private fun setupDetailsOverviewRowPresenter() {
        // Set detail background.
        val detailsPresenter = object : FullWidthDetailsOverviewRowPresenter(WorkDetailsDescriptionRender()) {

            private var mDetailsImageView: ImageView? = null

            override fun createRowViewHolder(parent: ViewGroup): RowPresenter.ViewHolder {
                val viewHolder = super.createRowViewHolder(parent)
                mDetailsImageView = viewHolder.view.findViewById(R.id.details_overview_image)
                val lp = mDetailsImageView?.layoutParams
                lp?.width = resources.getDimensionPixelSize(R.dimen.movie_width)
                lp?.height = resources.getDimensionPixelSize(R.dimen.movie_height)
                mDetailsImageView?.layoutParams = lp
                return viewHolder
            }
        }
        detailsPresenter.actionsBackgroundColor = resources.getColor(R.color.detail_view_actionbar_background, activity!!.theme)
        detailsPresenter.backgroundColor = resources.getColor(R.color.detail_view_background, activity!!.theme)

        // Hook up transition element.
        val sharedElementHelper = FullWidthDetailsOverviewSharedElementHelper()
        sharedElementHelper.setSharedElementEnterTransition(activity, SHARED_ELEMENT_NAME)
        detailsPresenter.setListener(sharedElementHelper)
        detailsPresenter.isParticipatingEntranceTransition = true
        detailsPresenter.setOnActionClickedListener { action ->
            var position = 0
            when (action.id.toInt()) {
                ACTION_FAVORITE -> presenter.setFavorite(work)
                ACTION_SIMILAR -> {
                    if (similarRowAdapter.size() > 0) {
                        position++
                    }
                    if (recommendedRowAdapter.size() > 0) {
                        position++
                    }
                    if (castRowAdapter.size() > 0) {
                        position++
                    }
                    if (videoRowAdapter.size() > 0) {
                        position++
                    }
                    setSelectedPosition(position)
                }
                ACTION_RECOMMENDED -> {
                    if (recommendedRowAdapter.size() > 0) {
                        position++
                    }
                    if (castRowAdapter.size() > 0) {
                        position++
                    }
                    if (videoRowAdapter.size() > 0) {
                        position++
                    }
                    setSelectedPosition(position)
                }
                ACTION_CAST -> {
                    if (castRowAdapter.size() > 0) {
                        position++
                    }
                    if (videoRowAdapter.size() > 0) {
                        position++
                    }
                    setSelectedPosition(position)
                }
                ACTION_VIDEOS -> {
                    if (videoRowAdapter.size() > 0) {
                        position++
                    }
                    setSelectedPosition(position)
                }
            }
        }
        presenterSelector.addClassPresenter(DetailsOverviewRow::class.java, detailsPresenter)
    }

    private inner class ItemViewSelectedListener : OnItemViewSelectedListener {

        override fun onItemSelected(viewHolder: Presenter.ViewHolder?, item: Any?, rowViewHolder: RowPresenter.ViewHolder?, row: Row?) {
            when (row?.headerItem?.id?.toInt()) {
                RECOMMENDED_HEADER_ID -> {
                    item?.let {
                        if (recommendedRowAdapter.indexOf(it) >= recommendedRowAdapter.size() - 1) {
                            presenter.loadRecommendationByWork(work)
                        }
                    }
                }
                SIMILAR_HEADER_ID -> {
                    item?.let {
                        if (similarRowAdapter.indexOf(it) >= similarRowAdapter.size() - 1) {
                            presenter.loadSimilarByWork(work)
                        }
                    }
                }
            }
        }
    }

    private inner class ItemViewClickedListener : OnItemViewClickedListener {

        override fun onItemClicked(itemViewHolder: Presenter.ViewHolder, item: Any, rowViewHolder: RowPresenter.ViewHolder, row: Row?) {
            when (row?.headerItem?.id?.toInt()) {
                CAST_HEAD_ID -> {
                    val cast = item as Cast
                    val castBundle = ActivityOptionsCompat.makeSceneTransitionAnimation(
                            activity!!,
                            (itemViewHolder.view as ImageCardView).mainImageView,
                            CastDetailsFragment.SHARED_ELEMENT_NAME
                    ).toBundle()
                    startActivity(CastDetailsActivity.newInstance(context, cast), castBundle)
                }
                VIDEO_HEADER_ID -> {
                    val video = item as Video
                    val intent = Intent(
                            Intent.ACTION_VIEW,
                            Uri.parse(String.format(BuildConfig.YOUTUBE_BASE_URL, video.key))
                    )
                    try {
                        startActivity(intent)
                    } catch (e: ActivityNotFoundException) {
                        Timber.e(e, "Failed to play a video")
                    }
                }
                RECOMMENDED_HEADER_ID, SIMILAR_HEADER_ID -> {
                    val work = item as Work
                    val bundle = ActivityOptionsCompat.makeSceneTransitionAnimation(
                            activity!!,
                            (itemViewHolder.view as ImageCardView).mainImageView,
                            SHARED_ELEMENT_NAME
                    ).toBundle()
                    startActivity(WorkDetailsActivity.newInstance(context, work), bundle)
                }
            }
        }
    }

    companion object {

        const val SHARED_ELEMENT_NAME = "SHARED_ELEMENT_NAME"
        const val WORK = "WORK"

        private const val ACTION_FAVORITE = 1
        private const val ACTION_VIDEOS = 2
        private const val ACTION_CAST = 3
        private const val ACTION_RECOMMENDED = 4
        private const val ACTION_SIMILAR = 5
        private const val VIDEO_HEADER_ID = 1
        private const val RECOMMENDED_HEADER_ID = 2
        private const val SIMILAR_HEADER_ID = 3
        private const val CAST_HEAD_ID = 4

        fun newInstance() = WorkDetailsFragment()
    }
}